### PR TITLE
Changed Logger

### PR DESCRIPTION
**Pull request by mvegter**

### CODE REVIEW

**Comment by Theo47** [comment]
It's getting late so i'll check this tomorrow.

**Comment by Theo47** [comment]
Did some minor changes that i thought were needed.

**Comment by mvegter** [comment]
@Theo47 You should fix the License in Helpers/Logger.cs -> Lib/AppLogger.cs as the change you made, should be under the modified work copyright.

**Comment by Theo47** [comment]
Right, i assume you mean that since i copied the WriteObject function from the old AppLogger, i should mention Rallion on the copyright.

How does it work in this case? Do i copy the original "Copyright 2011, 2012, 2013 Steve Labbe" notice to the new file?

**Comment by mvegter** [comment]
@Theo47 Just by renaming my methods you had to change the header to original work copyright martijn vegter, and then add yourself to the modified work copyright. Then you can add an auther tag in the writeobject method with Steve's name.

**Comment by Theo47** [comment]
I'll check it more tomorrow but i think claiming copyright for my modifications is optional.

https://www.gnu.org/licenses/gpl-faq.html#RequiredToClaimCopyright

**Comment by mvegter** [comment]
`To “modify” a work means to copy from or adapt all or part of the work in a fashion requiring copyright permission, other than the making of an exact copy. The resulting work is called a “modified version” of the earlier work or a work “based on” the earlier work.`

`For the developers' and authors' protection, the GPL clearly explains that there is no warranty for this free software. For both users' and authors' sake, the GPL requires that modified versions be marked as changed, so that their problems will not be attributed erroneously to authors of previous versions.`

So in short (this is no legal advice), because you changed the code I made, you have to clearly state it has been changed from the original.

**Comment by Theo47** [comment]
`A “covered work” means either the unmodified Program or a work based on the Program.`

`You may convey a work based on the Program, or the modifications to produce it from the Program, in the form of source code under the terms of section 4, provided that you also meet all of these conditions:`
`a) The work must carry prominent notices stating that you modified it, and giving a relevant date.`

The way i read it, work refers to the program as a whole rather than each individual file or function.

So since i modified Depressurizer, i should put a notice in a prominent place saying something like "This is a modified version of Depressurizer" (plus date).

The most prominent place i can think of is at the top of the readme.md file. So maybe change the existing "This is a continuation of rallion's Depressurizer." to a more explicit "This is a modified version of rallion's Depressurizer"?

p.s: Also not a lawyer.

**Comment by mvegter** [comment]
`To do so, attach the following notices to the program. It is safest to attach them to the start of each source file to most effectively state the exclusion of warranty; and each file should have at least the “copyright” line and a pointer to where the full notice is found.`

It states per file the exclusion per warranty including the copyright owner, it is per-file based.

**Comment by Theo47** [comment]
I'd say you are confusing the copyright notice with the "The work must carry prominent notices stating that you modified it". Those are two different things.

https://www.gnu.org/licenses/gpl-faq.html#NoticeInSourceFile
`Adding a copyright notice and a license notice at the start of each source file is easy and makes such confusion unlikely.`

Each file should have a copyright notice and a license notice. (But not necessarily a copyright notice for each contributor)

The work must carry prominent notices stating that you modified it only when you are conveying modified versions of the Program.

**Comment by mvegter** [comment]
So you are saying that if you make any type of change to a car (e.g. wheels, custom frame or anything), that you can still sell and / or promote it as an original?

**Comment by Theo47** [comment]
I'd say you are getting off-topic, comparing apples with oranges. A car is not under a gnu license. We could have a theoretical conversation about what would happen if cars were under gnu license but like i said we'd be getting off-topic

From a software perspective, if i go to Depressurizer/Depressurizer and e.g change the skin (paint the car), i wouldn't say that is is a modified version of Depressurizer/Depressurizer. I would say it is Depressurizer/Depressurizer v0.7.5.0.

**Comment by mvegter** [comment]
I do not approve nor have I written the changes that you made to the code, yet you failed to change the license header to state that the original work (this file) was modified. 

The effect this has that all liability and copyright infringement is on me, while you changed the code (thus owning the copyright on the changes and being responsible for proper usage of the license), you don't have to claim them but the moment you pushed them you own the copyrigh on it:

`Code without an explicit license is protected by copyright and is by default All Rights Reserved. The person or people who wrote the code are protected as such. Any time you're using software you didn't write, licensing should be considered and abided.` ~ Brian Doll, GitHub's VP of Marketing.

And I will once again quote the license itself:

`Developers that use the GNU GPL protect your rights with two steps: (1) assert copyright on the software, and (2) offer you this License giving you legal permission to copy, distribute and/or modify it.`

`For the developers' and authors' protection, the GPL clearly explains that there is no warranty for this free software. For both users' and authors' sake, the GPL requires that modified versions be marked as changed, so that their problems will not be attributed erroneously to authors of previous versions.`

You are free to make change, but you have to mark to changes as a modified version of the original.

**Comment by Theo47** [comment]
What you are suggesting would quickly get ridiculous. What happens if you send another pr where you edit logger.cs and i edit that too in turn? Would the notice look like:
>Original work copyright 2017 mvegter
>Modified work copyright 2017 theo47
>Modified work of modified work copyright 2017 mvegter
>Modified work of modified work of modified work copyright 2017 theo47

As for the copyright notice, like it says below, you don't have to add a notice for everyone that contributed to the file.

`In that case, you should always include a copyright notice in the name of main copyright holder of the file. You can also include copyright notices for other copyright holders as well, and this is a good idea for those who have contributed a large amount and for those who specifically ask for notices in their names. (Sometimes the license on code that you copy in may require preserving certain copyright notices.) But you don’t have to include a notice for everyone who contributed to the file (which would be rather inconvenient). `

Also since you directly contributed code to the repository (via a pr), this case may fall under section 11 of the gpl rather than section 5.

`A “contributor” is a copyright holder who authorizes use under this License of the Program or a work on which the Program is based. The work thus licensed is called the contributor's “contributor version”.`

`A contributor's “essential patent claims” are all patent claims owned or controlled by the contributor, whether already acquired or hereafter acquired, that would be infringed by some manner, permitted by this License, of making, using, or selling its contributor version, but do not include claims that would be infringed only as a consequence of further modification of the contributor version. For purposes of this definition, “control” includes the right to grant patent sublicenses in a manner consistent with the requirements of this License.`

`Each contributor grants you a non-exclusive, worldwide, royalty-free patent license under the contributor's essential patent claims, to make, use, sell, offer for sale, import and otherwise run, modify and propagate the contents of its contributor version.`

**Comment by mvegter** [comment]
Let's analyze what you quoted:

`A “contributor` (me) `is a copyright holder who authorizes use under this License` (GNU GPLv3) ` of the Program` (refers to any copyrightable work licensed under this License.) `or a work on which the Program is based` (A “covered work”) `. The work thus licensed is called the contributor's “contributor version”.`

`A contributor's “essential patent claims” are all patent claims owned or controlled by the contributor, whether already acquired or hereafter acquired, that would be infringed by some manner, permitted by this License, of making, using, or selling its contributor version, but do not include claims that would be infringed only as a consequence of further modification of the contributor version` (Thus a modified version of the Program) ` . For purposes of this definition, “control” includes the right to grant patent sublicenses in a manner consistent with the requirements of this License.`

`Each contributor grants you a non-exclusive ` (Any kind of usage, consistent with the requirements of this License.) `, worldwide` (non-region exclusive)`, royalty-free patent` (No fees or any cost at all) ` license under the contributor's essential patent claims, to make, use, sell, offer for sale, import and otherwise run, modify and propagate the contents of its contributor version.`

Thus the code added by a contributor is refered to as the Program (`refers to any copyrightable work licensed under this License. Each licensee is addressed as “you”. “Licensees” and “recipients” may be individuals or organizations.`), meaning that if you modify it (`the Program`) that you have to clearly state it is a modified version (`For the developers' and authors' protection, the GPL clearly explains that there is no warranty for this free software. For both users' and authors' sake, the GPL requires that modified versions be marked as changed, so that their problems will not be attributed erroneously to authors of previous versions.`).

**Comment by Theo47** [comment]
On a re-read, i think that section had to do with patents specifically and the word "contributor" was just misleading.

Anyway i send an email to licensing@fsf.org asking about this. I'll let you know once i get a response.

**Comment by mvegter** [comment]
What did you ask specifically? And have you received an anwser yet?

**Comment by Theo47** [comment]
>Under section 5 of GPL 3 it says:
>
>>You may convey a work based on the Program, or the modifications to produce it from the Program, in the form of source code under the terms of section 4, provided that you also meet all of these conditions:
>>
>>a) The work must carry prominent notices stating that you modified it, and giving a relevant date.
>
>
>For simplicity, i'll refer to "prominent notices stating that you modified it" as "modification notices" below.
>
>1. If someone sends me a pull request (on github), are they required to add modification notices to it? Or is it self-evident that a pull request would contain modified source code?
>2. If i accept the pull request and then edit some of the code they contributed, am i required to add modification notices?
>3. Can you give an example of a modification notice? Do i add it on the top of each file i modified or on a single prominent place inside the program?
>4. Would the commit history on github count as modification notices?

Haven't gotten an answer yet. Like i said, i'll let you know once i get an answer.

**Comment by mvegter** [comment]
@Theo47 Any news yet?

**Comment by Theo47** [comment]
No, nothing yet.

**Comment by mvegter** [comment]
@Theo47 This is taking too long for me. Either you fix the license as I stated, or revert the work I added. If you fail to do either one of these within the next 48 hours (thus untill friday 8th September 2017, 21:45 UTC+01:00), then I will follow with legal options through GitHub itself.

**Comment by Theo47** [comment]
I went ahead and reverted like you requested. 6d850f4cdf98dcd85a4ccb8115c67c0d9075fe64

**Comment by mvegter** [comment]
Shame you had revert it, but that choice is up to you. Of course implementing the same logic I provided in any commited work or pull request that has been or will be reverted, still falls under my copyright. 

Same principle goes for previously and future committed work. Any miss usage of the license (my interpretation) will be reported.
